### ISSUE TITLE
[Web] There is no obvious improvement when xnnpack turns on multithreading

### ISSUE THREAD

**Issue by Disliketalking** [microsoft/onnxruntime#13819]
### Describe the issue

There is no obvious improvement when xnnpack turns on multithreading.
Under the same number of threads, xnnpack will not be better than wasm.

I'm not sure if this has something to do with 
kOrtSessionOptionsConfigAllowIntraOpSpinning 
and
 intra_op_num_threads

### To reproduce

executionProviders: ['xnnpack'], 
ort.env.wasm.numThreads = 4;

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

master

### Execution Provider

Other / Unknown

**Comment by vacing** [comment]
The data format of the model is NCHW, but xnnpack prefers NHWC, it seems there are many transpose ops in profiling file.
I noticed that ort team add supports to NHWC ops, does it means ort tend to support both formats?
Hope some advices can be given on how to use xnnpack provider ^^ @ScottMcEleney

**Comment by vacing** [comment]
> As for `kOrtSessionOptionsConfigAllowIntraOpSpinning`, I would suggest turning off it by sessionOptions. Otherwise, the two threadpool inside will content and affect eathother.

when multi-thread enabled, ort will turn off session thread poll, and use global thread poll instead, so the options set in session option will not work.

**Comment by fs-eire** [comment]
> > As for `kOrtSessionOptionsConfigAllowIntraOpSpinning`, I would suggest turning off it by sessionOptions. Otherwise, the two threadpool inside will content and affect eathother.
> 
> when multi-thread enabled in wasm, ort will turn off session thread poll, and use global thread poll instead, so the options set in session option will not work.
> 
> ```c++
> #if defined(__EMSCRIPTEN_PTHREADS__)
>   RETURN_NULLPTR_IF_ERROR(DisablePerSessionThreads, session_options);
> #else
>   // must disable thread pool when WebAssembly multi-threads support is disabled.
>   RETURN_NULLPTR_IF_ERROR(SetIntraOpNumThreads, session_options, 1);
>   RETURN_NULLPTR_IF_ERROR(SetSessionExecutionMode, session_options, ORT_SEQUENTIAL);
> #endif
> ```

yes. for ort-web, the global threadpool is used and the per-session threadpool is always disabled. i am curious about whether xnnpack threadpool is created and initialized correctly in wasm.

**Comment by wejoncy** [comment]
Just discussd with @fs-eire, there is not a easy way to support muti-threading for XNNPACK EP currently. we will support it in the next step and keep here up-to-date of the latest progress.

**Comment by vacing** [comment]
> for ort-web, the global threadpool is used and the per-session threadpool is always disabled

could you give more detail about this, why force using global threadpool in ort-web?

**Comment by fs-eire** [comment]
> > for ort-web, the global threadpool is used and the per-session threadpool is always disabled
> 
> could you give more detail about this, why force using global threadpool in ort-web?

By default, the CPU threadpool is created when the InferenceSession initialize a model. When we work on the first version of ort-web we tested the webassembly threadpool and found it not working correctly - the reason was because the threadpool is not yet ready when we try to use it, and the solution is to initialize the threadpool earlier - so we create a global threadpool when initialize the OrtEnv.

**Comment by vacing** [comment]
> the solution is to initialize the threadpool earlier

does it mean we should reuse global threadpool in xnnpack too or create xnn threadpoll when initialize the OrtEnv?

**Comment by vacing** [comment]
> we will support it in the next step

is there a schedule or a plan about this, maybe we can paticipate in fixing this issue if you have more important features to develop

**Comment by wejoncy** [comment]
> > we will support it in the next step
> 
> is there a schedule or a plan about this, maybe we can paticipate in fixing this issue if you have more important features to develop

Your contribution is highly appreciated. For now, ORT-web requires muti-sesions share the global threadpool. However xnnpack expects its own threadpool and it's not global and each session will own a threadpool.

Besides, we need a mechenisim to set the spining feature and set intra_thread_nums for xnnpack's threadpool.

The last one is we also need to initialize xnnpack's threadpool at the begining in when we do for ort's global threadpool.

we need to figure out a appropriate solution for that.

**Comment by fs-eire** [comment]
the ort-web team will discuss about this. I consider "enabling multi-threading xnnpack EP on web" a whole task, which may include several technical items including this one.

your help is definitely welcome and appreciated. the basic idea is:
1. in `onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc`:
     - use `#if defined(__EMSCRIPTEN_PTHREADS__)` (this macro is only defined when building multi-thread webassembly) to guard code for per-EP threadpool. For multi-thread webassembly, change the code to use a global threadpool; for other build config we keep the existing behavior
     - extract calling to `pthreadpool_create()` to a global or static member function
3. in `onnxruntime/wasm/api.cc`, function `OrtInit()` implementation:
     - add API call to `SetGlobalSpinControl()` ( the equivalent of `kOrtSessionOptionsConfigAllowIntraOpSpinning` applying to global thread pool
     - call the exposed function above to create a threadpool for xnnpack. should only do this when multi-threading + xnnpack EP is specified.

**Comment by vacing** [comment]
> the ort-web team will discuss about this. I consider "enabling multi-threading xnnpack EP on web" a whole task, which may include several technical items including this one.
> 
> your help is definitely welcome and appreciated. the basic idea is:
> 
> 1. in `onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc`:
>    
>    * use `#if defined(__EMSCRIPTEN_PTHREADS__)` (this macro is only defined when building multi-thread webassembly) to guard code for per-EP threadpool. For multi-thread webassembly, change the code to use a global threadpool; for other build config we keep the existing behavior
>    * extract calling to `pthreadpool_create()` to a global or static member function
> 2. in `onnxruntime/wasm/api.cc`, function `OrtInit()` implementation:
>    
>    * add API call to `SetGlobalSpinControl()` ( the equivalent of `kOrtSessionOptionsConfigAllowIntraOpSpinning` applying to global thread pool
>    * call the exposed function above to create a threadpool for xnnpack. should only do this when multi-threading + xnnpack EP is specified.

Awesome, get it !
By the way, how could we determine xnnpack multithread works well already, is there any any positive or negtive signs?

**Comment by Disliketalking** [comment]
> the ort-web team will discuss about this. I consider "enabling multi-threading xnnpack EP on web" a whole task, which may include several technical items including this one.
> 
> your help is definitely welcome and appreciated. the basic idea is:
> 
> 1. in `onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc`:
>    
>    * use `#if defined(__EMSCRIPTEN_PTHREADS__)` (this macro is only defined when building multi-thread webassembly) to guard code for per-EP threadpool. For multi-thread webassembly, change the code to use a global threadpool; for other build config we keep the existing behavior
>    * extract calling to `pthreadpool_create()` to a global or static member function
> 2. in `onnxruntime/wasm/api.cc`, function `OrtInit()` implementation:
>    
>    * add API call to `SetGlobalSpinControl()` ( the equivalent of `kOrtSessionOptionsConfigAllowIntraOpSpinning` applying to global thread pool
>    * call the exposed function above to create a threadpool for xnnpack. should only do this when multi-threading + xnnpack EP is specified.

We try to create a global thread pool for xnnpack in api.cc and pass it to xnnpack_execution_provider.cc through session_options.

When we call `pthreadpool_create()` in api.cc, the thread pool creation fails and returns nullptr.

It seems that calling `pthreadpool_create()` in xnnpack_execution_provider.cc will still return nullptr

**Comment by fs-eire** [comment]
did you use `--enable_wasm_threads` flag in your build command line?

**Comment by Disliketalking** [comment]
> 

Yes, we can be sure that we used the `--enable_wasm_threads` flag，and we can observe that the number in the global thread pool is not one

**Comment by guschmue** [comment]
multithreading for the xnnpack ep needs work
